[c10d] Introduce dist.get_local_rank, dist.get_global_rank and dist.get_global_ranks

kumpera opened the pull request:
Those functions enable membership introspection into a ProcessGroup. A common scenario
that needs this is library code that consumes a PG but doesn't create it, which means
it likely doesn't know the global ranks used to create it.

Translating from local to global is necessary when using c10d collectives like broadcast
so if your library code adopts the convention of using local rank 0, it needs
to the following:

```python
import torch.distributed as dist

my_pg: dist.ProcessGroup = ...

def my_library_bcast(tensor)
    dist.broadcast(tensor, src=dist.get_global_rank(my_pg, local_rank=0), my_pg)

```

This implements some of the helpers needed to implement the `clone` API from: https://github.com/pytorch/pytorch/issues/81291

kumpera committed:
[c10d] Introduce dist.get_local_rank, dist.get_global_rank and dist.get_global_ranks.

Those functions enable membership introspection into a ProcessGroup. A common scenario
that needs this is library code that consumes a PG but doesn't create it, which means
it likely doesn't know the global ranks used to create it.

Translating from local to global is necessary when using c10d collectives like broadcast
so if your library code adopts the convention of using local rank 0, it needs
to the following:

```python
import torch.distributed as dist

my_pg: dist.ProcessGroup = ...

def my_library_bcast(tensor)
    dist.broadcast(tensor, src=dist.get_global_rank(my_pg), my_pg)

```

kumpera commented on torch/distributed/distributed_c10d.py:
```diff
@@ -301,41 +301,67 @@ def _warn_not_in_group(op_name):
+    return group_ranks[global_rank]
 
-def _get_global_rank(group, group_rank):
+def get_global_rank(group: ProcessGroup, local_rank: int) -> int:
```
This was a typo on my end, the right snippet should be: 

```
dist.broadcast(tensor, src=dist.get_global_rank(my_pg, local_rank=0), my_pg)
```

I agree that collectives should take local ranks but this beyond the scope of this PR, to be addressed separatedly.

The switch local ranks is easy to justify by just looking at how every collective is implemented:
```python

#pseudo-code
def broadcast(tensor, src, pg): 
  return pg.broadcast(tensor, src=get_local_rank(pg, src), pg)

```

xunnanxu replied:
Cool that makes sense.
We can tackle the api improvement in a separate PR.

kumpera commented on torch/distributed/distributed_c10d.py:
```diff
@@ -301,41 +301,67 @@ def _warn_not_in_group(op_name):
-    raise RuntimeError("The group rank is not part of the group")
+    raise RuntimeError(f"Global rank {local_rank} is not part of group {group}")
 
+def get_global_ranks(group: ProcessGroup) -> Sequence[int]:
```
This is the membership list of all ranks that are part of a PG translated to global ranks. It's the equivalent of the following:

```python
def get_global_ranks(pg):
  return [get_global_rank(pg, i) for i in range(0, pg.get_size())]
```

It essentially recovers the `ranks` argument passed to `new_group`.
Is the naming is too similar to `get_global_rank` that makes it confusing?

xunnanxu replied:
I see. I think the confusing part is nothing is using this API in this PR just yet. Some documentation would definitely help.

On the other hand, is this actually helpful by itself? What I mean is a user still cannot create a copy of a pg as input without knowing other params like timeout and pg backend type. I feel like in practice this particular call may be only useful for that purpose (for most cases I'd imagine `get_global_rank()` is more often used), and hence it might make more sense to offer a high level api like `clone()` (with signature similar to `new_group()`), unless you plan to introduce `get_timeout(pg)`, `get_backend(pg)` which seem a bit awkward.

rohan-varma replied:
Based on this discussion, it sounds like https://github.com/pytorch/pytorch/issues/58015 might be useful? i.e. to address the following concern:

> What I mean is a user still cannot create a copy of a pg as input without knowing other params like timeout and pg backend type

kumpera replied:
I believe offering introspection into framework objects to be really important regardless of specific usage.

While this API is not enough to implement `clone_group` it's a step in that direction. I plan to fill in the remaining gaps in follow-up PRs. Maybe I should have made this explicit from the beginning.

kumpera replied:
@rohan-varma my plan is to implement [#81291 ](https://github.com/pytorch/pytorch/issues/81291) design of `new_sub_group` and `clone`.

kumpera commented:
@pytorchmergebot rebase

rohan-varma commented on test/distributed/test_c10d_common.py:
```diff
@@ -1116,6 +1116,32 @@ def _test_warn_not_in_group(self, backend):
+        with self.assertRaisesRegex(RuntimeError, "not registered"):
+            dist.get_local_rank(DummyProcessGroup(self.rank, self.world_size), 0)
+
+        self.assertEqual(dist.get_global_rank(group, 0), 1)
```
Hmm, this is honestly a bit confusing for me, because `get_local_rank` seems to take in a global rank, and `get_global_rank` seems to take in the group's local rank. 

Also, one question around the semantics / definition of "local rank". Does this mean the relative rank within the node? 

i.e. what if I initialize

```
# assume global ranks contain 1-16 across 2 nodes

group = dist.new_group(ranks=range(1, 16))
```

which spans devices across 2 nodes.

What would `dist.get_local_rank(group, 9)` return?

kumpera replied:
The naming is confusing because some times local rank is used in reference to the host relative rank.
Perhaps it should be renamed to group rank?

It is confusing that `get_global_rank` takes a local rank and vice-versa, but that's the nature of this beast.

A clearer example:

```
pg = dist.new_group(ranks=[1, 5, 10, 20])

assert pg.get_size() == 4

# global rank 1 is local rank 0
# global rank 5 is local rank 1
# ...

assert dist.get_local_rank(pg, 5) == 1
assert dist.get_local_rank(pg, 20) == 3

assert dist.get_local_rank(pg, 5) # raises

# and the inverse
assert dist.get_global_rank(pg, 0) == 1
assert dist.get_global_rank(pg, 1) == 5
assert dist.get_global_rank(pg, 2) == 10
assert dist.get_global_rank(pg, 5) # raises
```

rohan-varma commented on test/distributed/test_c10d_common.py:
```diff
@@ -1116,6 +1116,32 @@ def _test_warn_not_in_group(self, backend):
             dist.barrier(group=group)
             dist.broadcast(x, src=0, group=group)
 
+    def _test_rank_membership(self, backend):
```
Can we do anything to simulate behavior in a multi-node scenario?

kumpera replied:
This change is not multi-node related, I guess this confusion is the same as the previous one.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -301,41 +301,67 @@ def _warn_not_in_group(op_name):
-            f"The global rank {rank} is not part of the group {group}"
-        ) from None
-    return group_rank
+        raise RuntimeError("Group {group} is not registered with torch.distributed")
```
"not registered" might be unclear to the end user. How about

"group not registered, please create group with dist.new_group API"

kumpera replied:
loved it

kumpera commented on torch/distributed/distributed_c10d.py:
```diff
@@ -301,41 +301,67 @@ def _warn_not_in_group(op_name):
-    raise RuntimeError("The group rank is not part of the group")
+    raise RuntimeError(f"Global rank {local_rank} is not part of group {group}")
 
+def get_global_ranks(group: ProcessGroup) -> Sequence[int]:
```
It can be used to create a new process group based on an existing one.
For example:
```
dist.new_group(ranks=dist.get_group_ranks(nccl_pg), backend="gloo")
```

Right now this PR doesn't provide all accessors needed for this API to be general purpose enough, one can't access timeouts and options yet.

kumpera commented on torch/distributed/distributed_c10d.py:
```diff
@@ -301,41 +301,67 @@ def _warn_not_in_group(op_name):
-        if grp_rank == group_rank:
+        return local_rank
+    if group not in _pg_group_ranks:
+        raise RuntimeError("Group {group} is not registered with torch.distributed")
```
fixed

kumpera commented on torch/distributed/distributed_c10d.py:
```diff
@@ -301,41 +301,67 @@ def _warn_not_in_group(op_name):
-            f"The global rank {rank} is not part of the group {group}"
-        ) from None
-    return group_rank
+        raise RuntimeError("Group {group} is not registered with torch.distributed")
```
fixed.

kumpera commented:
This PR is blocked on us agreeing on whether to use the term local rank or group rank.

kumpera committed:
fix error messages.

rohan-varma commented:
@kumpera thanks for the explanation, `group_rank` sounds good.

kumpera committed:
Rename local rank to group rank.

kumpera committed:
Rename get_global_ranks to get_process_group_ranks

kumpera commented:
I made the following changes to address all feedback:

Renamed local rank to group rank.
Renamed `get_group_ranks` to `get_process_group_ranks` to make it less ambiguous.

As for use cases for `get_process_group_ranks`, it turns out this is useful in our FSDP +TP work when generating chunking specs and computing the device number. Additionally, the work to support checkpointing will need it as well.

kwen2501 commented:
Sorry for coming to the party late, and thanks for adding these APIs.

Curious: why is `get_group_ranks` ambiguous? Is it because it looks too similar to `get_group_rank`? If that's the case, would `get_all_ranks(group)` help?

Also, for this API, do we want to also support the default PG? i.e. group=None. User can probably get the list of ranks for the default PG in other ways, but since the default PG is also a valid PG, hence my thought here. (Also because traditional `get_rank(group)` API accepts None as input argument.)

kumpera commented:
@pytorchmergebot rebase

kumpera commented:
> Sorry for coming to the party late, and thanks for adding these APIs.
> 
> Curious: why is `get_group_ranks` ambiguous? Is it because it looks too similar to `get_group_rank`? If that's the case, would `get_all_ranks(group)` help?

Yes, I think the name was too similar and not clear about it.
> 
> Also, for this API, do we want to also support the default PG? i.e. group=None. User can probably get the list of ranks for the default PG in other ways, but since the default PG is also a valid PG, hence my thought here. (Also because traditional `get_rank(group)` API accepts None as input argument.)

The 3 new functions support the default group if you pass it explicitly.
It sort of makes sense to support `None` as a valid value for consistency sake.

kumpera commented:
@pytorchmergebot  rebase

rohan-varma commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -2747,6 +2747,11 @@ def test_nccl_warn_not_in_group_debug_info(self):
+    @requires_nccl()
+    @skip_if_lt_x_gpu(2)
+    def test_nncl_rank_membership(self):
+        self._test_rank_membership(backend="gloo")
```
backend="nccl"?

kumpera replied:
🤦

rohan-varma reviewed (approved):
APIs look good, just one minor comment on the test

kumpera committed:
Fix test_nncl_rank_membership test

kumpera commented:
@pytorchmergebot rebase

kumpera commented:
@pytorchmergebot  merge

kumpera committed:
Worsen typing until we fix C10D public API.

kumpera committed:
fix test.

kumpera commented:
@pytorchmergebot  merge

kumpera committed:
Merge branch 'master' into pg_ext
